src/core/src/sketch/minhash.rs

```diff
@@ -458,11 +458,38 @@ impl KmerMinHash {
         Ok(())
     }
 
-    pub fn count_common(&self, other: &KmerMinHash) -> Result<u64, Error> {
-        self.check_compatible(other)?;
-        let iter = Intersection::new(self.mins.iter(), other.mins.iter());
+    pub fn count_common(&self, other: &KmerMinHash, downsample: bool) -> Result<u64, Error> {
+        if downsample && self.max_hash != other.max_hash {
+            if self.max_hash < other.max_hash {
+                let mut new_mh = KmerMinHash::new(
+                    other.num,
+                    other.ksize,
+                    other.hash_function,
+                    other.seed,
+                    self.max_hash,
+                    other.abunds.is_some(),
+                );
+                new_mh.add_many(&other.mins)?;
+                self.count_common(&new_mh, false)
```

Comment:
This if clause and the else below are very similar. If you wanted to, you could try to DRY up the code. Something along this might work:

```rust
let cmp = self.max_hash < other.max_hash;
let a = if cmp { self } else { other };
let b = if !cmp { self } else { other };

let mut new_mh = KmerMinHash::new(
    b.num,
    b.ksize,
    b.hash_function,
    b.seed,
    a.max_hash,
    b.abunds.is_some(),
);
new_mh.add_many(&b.mins)?;
new_mh.count_common(a, false)
```

Maybe you have to drop some mutable references in there. Just an idea.